lib/linguist/heuristics.rb

```diff
@@ -204,11 +204,36 @@ def call(data)
     end
 
     disambiguate ".h" do |data|
+     CPPRegex = /
+  ^ \s*
+  (?: template \s* <
+    | \#.*include.*<
+      (stdint|cstdint|string|vector|map|list
+      |array|bitset|queue|stack|forward_list
+      |unordered_(map|set)|(i|o|io)stream).*>
+    )
+  |
+  ^ [\x20\t]*
+  (?: try \b
+    | catch \s *\(
+    | ( class
+      | namespace
+      | using \h+ namespace
+      )
+      \s+ \w+
+    | (private|public|protected) :$
+    )
+  | std::\w+
+  | .*\#ifdef.*__cplusplus.*
+  | ^struct\s+
+/xi unless const_defined?(:CPPRegex)
+
       if ObjectiveCRegex.match(data)
         Language["Objective-C"]
-      elsif (/^\s*#\s*include <(cstdint|string|vector|map|list|array|bitset|queue|stack|forward_list|unordered_map|unordered_set|(i|o|io)stream)>/.match(data) ||
-        /^\s*template\s*</.match(data) || /^[ \t]*try/.match(data) || /^[ \t]*catch\s*\(/.match(data) || /^[ \t]*(class|(using[ \t]+)?namespace)\s+\w+/.match(data) || /^[ \t]*(private|public|protected):$/.match(data) || /std::\w+/.match(data))
-        Language["C++"]
+      elsif CPPRegex.match(data)
+           Language["C++"]
+      else
+        Language["C"]
```

Comment:
That's going to classify way too many files as C. I'd prefer that we let the classifier decide, even if it is a bit arbitrary for now.